### PR TITLE
INFRA-935: Add a limit to collection-label association

### PR DESCRIPTION
**Pull request by katerinachinnappan**
## Goal

Currently, a collection can have a max of 20 labels. Add this limit to `createCollection` and `updateCollection` mutations.

## Todos

* [ ] After review, push to dev before merge.

## Tickets

* [https://getpocket.atlassian.net/browse/INFRA-935](https://getpocket.atlassian.net/browse/INFRA-935)

## Implementation Decisions

* Added a `collectionLabelLimit` config variable to `src/config` to store the collection-label association limit which can be easily updated in the future if needed.
* Added a `checkCollectionLabelLimit(labelExternalIds: string[])` to `src/database/utils.ts` utility function to check the limit.

### CODE REVIEW

File: src/database/mutations/Collection.ts [hunk 1]
```diff
@@ -82,7 +83,11 @@ export async function createCollection(
   }
 
   // if a collection has any labels, set up a connection to labels, too
+  // first check if collection - label limit is not exceeded
```
Comment (nina-py) [inline src/database/mutations/Collection.ts:86]: Nitpick: compound nouns as adjectives should be hyphenated, so "collection-label limit>

File: src/database/mutations/Collection.ts [hunk 2]
```diff
@@ -256,7 +261,11 @@ export async function updateCollection(
   });
 
   // If a collection has any labels, set up a connection to labels, too
+  // first check if collection - label limit is not exceeded
```
Comment (nina-py) [inline src/database/mutations/Collection.ts:264]: Same dash vs. hyphen nitpick as elsewhere

File: src/database/utils.ts [hunk 3]
```diff
@@ -69,3 +70,14 @@ export function buildSearchCollectionsWhereClause(
...
   return where;
 }
+
+export function checkCollectionLabelLimit(labelExternalIds: string[]) {
+  if (labelExternalIds.length > config.app.collectionLabelLimit) {
+    const labelsToRemove =
+      labelExternalIds.length - config.app.collectionLabelLimit;
+    const labelsSingularOrPlural = labelsToRemove === 1 ? 'label' : 'labels';
+    throw new UserInputError(
+      `Too many labels provided. Max is ${config.app.collectionLabelLimit}. Please remove ${labelsToRemove} ${labelsSingularOrPlural}.`
```
Comment (nina-py) [inline src/database/utils.ts:80]: I'm on the fence about the wording of this error message. It's on the long side and specifies the number of labels to remove as a number, not a word, which is awkward to read - there's an example in the tests. 

One way to improve the existing wording is to use `toWords()` to specify the number of labels to remove. 

Maybe the wording of the entire message could be reworked, I'm not sure. This is an edge case that will probably not be seen that often in real life, so I'm trying not to go overboard with suggestions for perfect wording. How about something along the lines of "Too many labels: 21 out of 20 max provided"? This would also eliminate the stylistic need for converting numbers into words! 🙃

**Review by nina-py — COMMENTED** [review]
Added a couple of nitpicks around error message and comment wording. Excellent work! 🚀

File: src/database/utils.ts (hunk 3 above)
Comment (katerinachinnappan) [reply src/database/utils.ts:80]: @nina-py mmm yes, I was thinking how to word this one. I do want to definitely provide the max allowed # and the users input in the message. I do think it would be useful to provide how many labels to remove --- say our max is 20 and someone goes overboard and uses 36 labels, it would help to show that they should remove 16 labels. Idk tbh. 

In this kind of use case it also makes me think we should have the functionality to remove the extra labels:
* User enters too many labels
* We show error message and option "remove N # of labels?" and if user clicks that, it would remove the last N extra labels. 

I see this being useful in cases where user has more than 5 extra labels for example. This would save time for them rather than clicking N # of times to delete the extra labels. 

Anyways this is getting out of the scope for this ticket, maybe we can reword the message to: "The number of provided labels must not be more than 20. It is currently 21".

Comment (nina-py) [reply src/database/utils.ts:80]: I think removing extra labels is fraught with danger as we wouldn't necessarily be removing the labels the user would have chosen to remove if there was a need to pare the number of labels down. The user would probably prioritise some labels over others instead of chopping off the end of the list which we'd do programmatically.

I really like the last wording you provided, I think it's the best of the lot. Short and has all the actionable information.

**Review by nina-py — APPROVED** [review]
